lib/ParserHelpers.js

```diff
@@ -83,3 +124,1129 @@ ParserHelpers.skipTraversal = function skipTraversal() {
+	// 	 * e.g.: new DefinePlugin({
+	// 	 * "typeof a": "tyepof b",
+	// 	 * "typeof b": "typeof a"
+	// 	 * });
+	// 	 */
+	var recurseTypeof = opts.recurseTypeof;
+	if(recurseTypeof) return;
+	recurseTypeof = true;
+	const res = parser.evaluate(opts.typeofCode);
+	recurseTypeof = false;
+	res.setRange(expr.range);
+	return res;
+};
+
+ParserHelpers.DefinePluginTypeof = function DefinePluginTypeof(expr, opts, parser) {
+	const res = parser.evaluate(opts.typeofCode);
+	if(!res.isString()) return;
+	return ParserHelpers.toConstantDependency(expr, {
+		code: JSON.stringify(res.string)
+	}, parser);
+};
+
+ParserHelpers.CommonJSPluginAssignRequire = function CommonJSPluginAssignRequire(expr, opts, parser) {
+	// to not leak to global "require", we need to define a local require here.
+	const dep = new ConstDependency("var require;", 0);
+	dep.loc = expr.loc;
+	parser.state.current.addDependency(dep);
+	parser.scope.definitions.push("require");
+	return true;
+};
+
+ParserHelpers.CommonJSPluginRenameRequire = function CommonJSPluginRenameRequire(expr, opts, parser) {
+	// define the require variable. It's still undefined, but not "not defined".
+	const dep = new ConstDependency("var require;", 0);
+	dep.loc = expr.loc;
+	parser.state.current.addDependency(dep);
+	return false;
+};
+
+ParserHelpers.CommonJSPluginExpressionRequire = function CommonJSPluginExpressionRequire(expr, opts, parser) {
+	const dep = new CommonJsRequireContextDependency(opts.unknownContextRequest, opts.unknownContextRecursive, opts.unknownContextRegExp, expr.range);
+	dep.critical = opts.unknownContextCritical && "require function is used in a way in which dependencies cannot be statically extracted";
+	dep.loc = expr.loc;
+	dep.optional = !!parser.scope.inTry;
+	parser.state.current.addDependency(dep);
+	return true;
+};
+
+ParserHelpers.CommonJSPluginCallRequire = function CommonJSPluginCallRequire(expr, opts, parser) {
+	if(expr.arguments.length !== 1) return;
+	let localModule;
+	const param = parser.evaluateExpression(expr.arguments[0]);
+	if(param.isConditional()) {
+		let isExpression = false;
+		const prevLength = parser.state.current.dependencies.length;
+		const dep = new RequireHeaderDependency(expr.callee.range);
+		dep.loc = expr.loc;
+		parser.state.current.addDependency(dep);
+		param.options.forEach(function(param) {
+			const result = parser.applyPluginsBailResult("call require:commonjs:item", expr, param);
+			if(result === undefined) {
+				isExpression = true;
+			}
+		});
+		if(isExpression) {
+			parser.state.current.dependencies.length = prevLength;
+		} else {
+			return true;
+		}
+	}
+	if(param.isString() && (localModule = LocalModulesHelpers.getLocalModule(parser.state, param.string))) {
+		const dep = new LocalModuleDependency(localModule, expr.range);
+		dep.loc = expr.loc;
+		parser.state.current.addDependency(dep);
+		return true;
+	} else {
+		const result = parser.applyPluginsBailResult("call require:commonjs:item", expr, param);
```

Comment:
I think that these are all specific enough that they don't need to be plugins... this should probably just be a function call.

Are parser plugins common outside of the webpack code base?

Reply:
Adding on this... by allowing plugins to create their own events we don't have a standard api for `parser.plugin` this makes refactoring much harder especially when stuff like `parser.plugin('call ' + expr, () => {})` are used.